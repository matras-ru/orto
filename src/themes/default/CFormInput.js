export const baseClass =
    'form-input-custom inline-block rounded-lg bg-white text-black-100 border-2 px-1-2 outline-none resize-none focus:border-primary-100 focus:text-black-100 transition-border transition-250 transition-ease-in-out';
export const defaultClass = 'border-primary-200';
export const defaultSizeClass = 'h-2-8';
export const largeSizeClass = 'h-12-4 py-0-6';
export const errorClass = 'form-input-is-error border-danger text-danger';
export const notEmptyClass = 'form-input-not-empty';
export const wrapperClass = 'relative inline-block';

export default {
    baseClass,
    defaultClass,
    defaultSizeClass,
    largeSizeClass,
    errorClass,
    notEmptyClass,
    wrapperClass
};
