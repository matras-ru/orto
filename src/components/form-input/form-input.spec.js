/* eslint-env jest */
import { mount, createLocalVue } from '@vue/test-utils';
import CFormInput from './form-input';

describe('Input basic', () => {
    it('is rendered', () => {
        const wrapper = mount(CFormInput, {
            propsData: {
                type: 'text',
                id: 'foo',
                name: 'foo'
            }
        });

        const input = wrapper.find('input');

        expect(wrapper.element.tagName).toEqual('LABEL');
        expect(input.element.tagName).toEqual('INPUT');
        expect(input.attributes('type')).toBeDefined();
        expect(input.attributes('type')).toBe('text');
        expect(input.attributes('id')).toBeDefined();
        expect(input.attributes('id')).toBe('foo');
        expect(input.attributes('name')).toBeDefined();
        expect(input.attributes('name')).toBe('foo');

        wrapper.destroy();
    });
});

describe('Input types', () => {
    it('textarea', () => {
        const wrapper = mount(CFormInput, {
            propsData: {
                type: 'textarea',
                id: 'foo',
                name: 'foo'
            }
        });

        const textarea = wrapper.find('textarea');

        expect(wrapper.element.tagName).toEqual('LABEL');
        expect(textarea.element.tagName).toEqual('TEXTAREA');
        expect(textarea.attributes('type')).not.toBeDefined();
        expect(textarea.attributes('id')).toBeDefined();
        expect(textarea.attributes('id')).toBe('foo');
        expect(textarea.attributes('name')).toBeDefined();
        expect(textarea.attributes('name')).toBe('foo');
        expect(textarea.classes()).toContain('resize-none');

        wrapper.destroy();
    });

    it('number', () => {
        const wrapper = mount(CFormInput, {
            propsData: {
                type: 'number',
                id: 'foo',
                name: 'foo'
            }
        });

        const input = wrapper.find('input');

        expect(input.attributes('type')).toBeDefined();
        expect(input.attributes('type')).toBe('number');

        wrapper.destroy();
    });

    it('number min/max', () => {
        const wrapper = mount(CFormInput, {
            propsData: {
                type: 'number',
                id: 'foo',
                name: 'foo',
                min: 0,
                max: 100
            }
        });

        const input = wrapper.find('input');

        expect(input.attributes('type')).toBeDefined();
        expect(input.attributes('type')).toBe('number');
        expect(input.attributes('min')).toBeDefined();
        expect(input.attributes('min')).toBe('0');
        expect(input.attributes('max')).toBeDefined();
        expect(input.attributes('max')).toBe('100');

        wrapper.destroy();
    });
});

describe('Input states', () => {
    let wrapper;
    let innerWrapper;
    let input;

    beforeEach(() => {
        wrapper = mount(CFormInput);
        input = wrapper.find('input');
        innerWrapper = wrapper.find('div');
    });

    afterEach(() => {
        input = null;
        innerWrapper = null;
        wrapper.destroy();
    });

    it('empty', async () => {
        wrapper.setProps({ label: 'label' });
        await wrapper.vm.$nextTick();
        const label = wrapper.findComponent({ ref: 'label' });
        expect(innerWrapper.classes('border-black-200')).toBe(true);

        expect(label.classes()).not.toContain('transform');
        expect(label.classes()).not.toContain('-translate-y-full');
        expect(label.classes()).not.toContain('scale-75');
    });

    it('empty label (show placeholder)', async () => {
        wrapper.setProps({ placeholder: 'label' });
        await wrapper.vm.$nextTick();

        const input = wrapper.find('input');
        expect(input.classes()).toContain('form-input-not-label');
    });

    it('not empty', async () => {
        wrapper.setProps({ modelValue: 'foo', label: 'label' });
        await wrapper.vm.$nextTick();
        const label = wrapper.findComponent({ ref: 'label' });

        expect(innerWrapper.classes('border-black-200')).toBe(true);
        expect(label.classes()).toContain('transform');
        expect(label.classes()).toContain('-translate-y-full');
        expect(label.classes()).toContain('scale-75');
    });

    it('error', async () => {
        wrapper.setProps({ error: true, label: 'label' });
        await wrapper.vm.$nextTick();
        const label = wrapper.findComponent({ ref: 'label' });

        expect(innerWrapper.classes('border-danger')).toBe(true);
        expect(label.classes('text-danger')).toBe(true);
    });

    it('focus', async () => {
        wrapper.setProps({ label: 'label' });
        input.trigger('focus');

        await wrapper.vm.$nextTick();
        const label = wrapper.findComponent({ ref: 'label' });

        expect(innerWrapper.classes('border-primary-100')).toBe(true);
        expect(label.classes()).toContain('transform');
        expect(label.classes()).toContain('-translate-y-full');
        expect(label.classes()).toContain('scale-75');
    });
});

describe('Input events', () => {
    it('emits an input event', () => {
        const wrapper = mount(CFormInput, {
            propsData: {
                type: 'text',
                id: 'foo',
                name: 'foo'
            }
        });
        const input = wrapper.find('input');

        input.element.value = 'test';
        input.trigger('input');

        expect(wrapper.emitted('input')).toBeDefined();
        expect(wrapper.emitted().input.length).toEqual(1);
        expect(wrapper.emitted().input[0][0]).toEqual('test');

        wrapper.destroy();
    });

    //

    it('v-model', () => {
        const localVue = new createLocalVue();

        const App = localVue.extend({
            data() {
                return {
                    model: ''
                };
            },

            render(h) {
                return h(CFormInput, {
                    props: {
                        type: 'text',
                        id: 'foo',
                        name: 'foo',
                        modelValue: this.model
                    },
                    on: {
                        input: val => {
                            this.model = val;
                        }
                    }
                });
            }
        });

        const wrapper = mount(App, {
            localVue: localVue
        });

        const input = wrapper.find('input');

        expect(wrapper.vm.model).toEqual('');
        input.setValue('test');
        expect(wrapper.vm.model).toEqual('test');

        wrapper.destroy();
    });
});
