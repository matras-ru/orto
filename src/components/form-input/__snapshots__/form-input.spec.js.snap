// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input basic is rendered 1`] = `
<label class="block px-0-4 mb-0-8">
  <div class="flex items-center border-2 rounded-lg -mx-0-4 transition-border duration-150 border-black-200 px-0-8">
    <div class="flex-auto relative py-0-5">
      <div class="relative"><input name="foo" id="foo" type="text" class="w-full form-input form-input-not-label"></div>
    </div>
  </div>
</label>
`;

exports[`Input states empty 1`] = `
<label class="block px-0-4 mb-0-8">
  <div class="flex items-center border-2 rounded-lg -mx-0-4 transition-border duration-150 border-black-200 px-0-8">
    <div class="flex-auto relative py-0-5">
      <div class="relative"><input type="text" class="w-full form-input"></div>
      <div class="absolute left-0 max-w-full truncate pointer-events-none uppercase origin-top-left transition-transform ease-in duration-150 leading-snug bg-white text-tertiary-300 px-0-4 top-0-5">label</div>
    </div>
  </div>
</label>
`;

exports[`Input states empty label (show placeholder) 1`] = `
<label class="block px-0-4 mb-0-8">
  <div class="flex items-center border-2 rounded-lg -mx-0-4 transition-border duration-150 border-black-200 px-0-8">
    <div class="flex-auto relative py-0-5">
      <div class="relative"><input type="text" class="w-full form-input form-input-not-label" placeholder="label"></div>
    </div>
  </div>
</label>
`;

exports[`Input states error 1`] = `
<label class="block px-0-4 mb-0-8">
  <div class="flex items-center border-2 rounded-lg -mx-0-4 transition-border duration-150 border-danger px-0-8">
    <div class="flex-auto relative py-0-5">
      <div class="relative"><input type="text" class="w-full form-input"></div>
      <div class="absolute left-0 max-w-full truncate pointer-events-none uppercase origin-top-left transition-transform ease-in duration-150 leading-snug bg-white text-danger px-0-4 top-0-5">label</div>
    </div>
  </div>
</label>
`;

exports[`Input states not empty 1`] = `
<label class="block px-0-4 mb-0-8">
  <div class="flex items-center border-2 rounded-lg -mx-0-4 transition-border duration-150 border-black-200 px-0-8">
    <div class="flex-auto relative py-0-5">
      <div class="relative"><input type="text" class="w-full form-input"></div>
      <div class="absolute left-0 max-w-full truncate pointer-events-none uppercase origin-top-left transition-transform ease-in duration-150 leading-snug bg-white transform -translate-y-full scale-75 text-tertiary-300 px-0-4 top-0-5">label</div>
    </div>
  </div>
</label>
`;

exports[`Input types number 1`] = `
<label class="block px-0-4 mb-0-8">
  <div class="flex items-center border-2 rounded-lg -mx-0-4 transition-border duration-150 border-black-200 px-0-8">
    <div class="flex-auto relative py-0-5">
      <div class="relative"><input name="foo" id="foo" type="number" class="w-full form-input form-input-not-label">
        <div class="absolute top-0 right-0 flex flex-col h-full"><span class="inline-flex items-center justify-center h-1/2 cursor-pointer select-none opacity-50 pointer-events-none"></span><span class="inline-flex items-center justify-center h-1/2 cursor-pointer select-none opacity-50 pointer-events-none"></span></div>
      </div>
    </div>
  </div>
</label>
`;

exports[`Input types number min/max 1`] = `
<label class="block px-0-4 mb-0-8">
  <div class="flex items-center border-2 rounded-lg -mx-0-4 transition-border duration-150 border-black-200 px-0-8">
    <div class="flex-auto relative py-0-5">
      <div class="relative"><input name="foo" id="foo" type="number" min="0" max="100" class="w-full form-input form-input-not-label">
        <div class="absolute top-0 right-0 flex flex-col h-full"><span class="inline-flex items-center justify-center h-1/2 cursor-pointer select-none"></span><span class="inline-flex items-center justify-center h-1/2 cursor-pointer select-none opacity-50 pointer-events-none"></span></div>
      </div>
    </div>
  </div>
</label>
`;

exports[`Input types textarea 1`] = `
<label class="block px-0-4 mb-0-8">
  <div class="flex items-center border-2 rounded-lg -mx-0-4 transition-border duration-150 border-black-200 px-0-8">
    <div class="flex-auto relative py-0-5">
      <div class="relative"><textarea name="foo" id="foo" rows="6" class="w-full form-input resize-none form-input-not-label"></textarea></div>
    </div>
  </div>
</label>
`;
